### PR TITLE
[Tests] Fix tests 3rd ed.

### PR DESCRIPTION
**Pull request by zhzhcookie**
### PR Category
OP Test

### Type of Change
Tests

### Description
1. Change --device=cpu to --ref=cpu
2. Add --mode=quick to test only one shape for CI cpu tests.
3. Add scalar shape for pointwise_dynamic op tests
4. Move assert_close and assert_equal to flag_gems.testing
5. Add mark for each op test, and register it to pytest.ini
6. Add speedup in benchmark print result.

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### CODE REVIEW

File: src/flag_gems/testing/__init__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    torch.float32: 1.3e-6,
+    torch.bfloat16: 0.016,
+}
+
+
+def assert_close(res, ref, dtype, equal_nan=False, reduce_dim=1):
+    ref = ref.to(dtype)
+    atol = 1e-4 * reduce_dim
+    rtol = RESOLUTION[dtype]
+    torch.testing.assert_close(res, ref, atol=atol, rtol=rtol, equal_nan=equal_nan)
```
Comment (tongxin) [inline src/flag_gems/testing/__init__.py]: Two issues:
- fp64 is not in RESOLUTION's key set.
- passing a dtype matching neither ref nor res may raise an error.

Comment (tongxin) [reply src/flag_gems/testing/__init__.py]: Also, do we consider default dtype=None? Now dtype is mandate.

File: tests/test_binary_pointwise_ops.py [hunk 2]
```diff
@@ -21,6 +21,11 @@
 from .conftest import TO_CPU
 
 
+def remove_zero(inp):
```
Comment (tongxin) [inline tests/test_binary_pointwise_ops.py]: I would rename this as replace_zeros

File: tests/accuracy_utils.py [hunk 3]
```diff
@@ -80,26 +75,30 @@ def to_reference(inp, upcast=False):
     if inp is None:
         return None
     ref_inp = inp
-    if TO_CPU:
-        ref_inp = ref_inp.to("cpu")
     if upcast:
         ref_inp = ref_inp.to(torch.float64)
```
Comment (tongxin) [inline tests/accuracy_utils.py:79]: what about integer? we are not supposed to upcast integers to floats in all occasions.

File: tests/accuracy_utils.py [hunk 4]
```diff
@@ -80,26 +75,30 @@ def to_reference(inp, upcast=False):
...
+        res = res.to("cpu")
+        assert ref.device == torch.device("cpu")
+    return res
 
 
-def gems_assert_equal(a, b):
-    if TO_CPU:
-        a = a.to("cpu")
-    assert torch.equal(a, b)
+def gems_assert_close(res, ref, dtype, equal_nan=False, reduce_dim=1):
```
Comment (tongxin) [inline tests/accuracy_utils.py:92]: what's different between this and testing.assert_close?

Comment (zhzhcookie) [reply tests/accuracy_utils.py:92]: > what's different between this and testing.assert_close?

TO_CPU here.

File: tests/test_binary_pointwise_ops.py (hunk 2 above)
Comment (zhzhcookie) [reply tests/test_binary_pointwise_ops.py]: > I would rename this as replace_zeros

done

File: tests/accuracy_utils.py (hunk 3 above)
Comment (zhzhcookie) [reply tests/accuracy_utils.py:79]: > what about integer? we are not supposed to upcast integers to floats in all occasions.

Now to_reference is only for float inputs.

File: src/flag_gems/testing/__init__.py (hunk 1 above)
Comment (zhzhcookie) [reply src/flag_gems/testing/__init__.py]: > Two issues:
> 
> * fp64 is not in RESOLUTION's key set.
> * passing a dtype matching neither ref nor res may raise an error.

Now only isclose & allclose test fp64. These two tests do not use assert_close.
Maybe we can assert res.dtype == dtype here?

Comment (zhzhcookie) [reply src/flag_gems/testing/__init__.py]: > > Two issues:
> > 
> > * fp64 is not in RESOLUTION's key set.
> > * passing a dtype matching neither ref nor res may raise an error.
> 
> Now only isclose & allclose test fp64. These two tests do not use assert_close. Maybe we can assert res.dtype == dtype here?

done

File: tools/op-unit-test.sh [hunk 5]
```diff
@@ -32,15 +32,16 @@ cmds=(
...
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_reduction_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_general_reduction_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_norm_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_unary_pointwise_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_binary_pointwise_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_special_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_pointwise_type_promotion.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_tensor_constructor_ops.py --ref=cpu --mode=quick && \
+    bash tools/pytest_mark_check.sh &"
```
Comment (Bowen12992) [inline tools/op-unit-test.sh]: it's better to put `pytest_mark_check.sh` before running the unit tests. Or use an additional ci pipeline named `pytest_mark_check` to run this check

File: tools/op-unit-test.sh [hunk 6]
```diff
@@ -32,15 +32,16 @@ cmds=(
...
+   "CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_blas_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_reduction_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_general_reduction_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_norm_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_unary_pointwise_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_binary_pointwise_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_special_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_pointwise_type_promotion.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_tensor_constructor_ops.py --ref=cpu --mode=quick && \
```
Comment (Bowen12992) [inline tools/op-unit-test.sh]:
```suggestion
    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_tensor_constructor_ops.py --ref=cpu --mode=quick &
```

**Review by Bowen12992 — COMMENTED** [review]
Add to [CONTRIBUTING.md](https://github.com/FlagOpen/FlagGems/blob/master/CONTRIBUTING.md)

File: tools/op-unit-test.sh [hunk 7]
```diff
@@ -32,15 +32,16 @@ cmds=(
...
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_reduction_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_general_reduction_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_norm_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_unary_pointwise_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_binary_pointwise_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_special_ops.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_pointwise_type_promotion.py --ref=cpu --mode=quick && \
+    CUDA_VISIBLE_DEVICES=7 coverage run ${COVERAGE_ARGS} -m pytest -s tests/test_tensor_constructor_ops.py --ref=cpu --mode=quick &"
+   "bash tools/pytest_mark_check.sh &"
```
Comment (Bowen12992) [inline tools/op-unit-test.sh]:
```
[pytest]
filterwarnings =
    ignore::pytest.PytestUnknownMarkWarning
```

Comment (zhzhcookie) [reply tools/op-unit-test.sh]: > ```
> [pytest]
> filterwarnings =
>     ignore::pytest.PytestUnknownMarkWarning
> ```

done

**Review by Bowen12992 — APPROVED** [review]
LGTM

**Pull request merged**